src/main/java/gg/projecteden/nexus/features/listeners/Sleep.java

```diff
@@ -16,115 +14,92 @@
+	private static final Map<World, State> sleepingWorlds = new HashMap<>();
+
+	static {
+		Tasks.repeatAsync(0, 1, () -> {
+			for (World world : sleepingWorlds.keySet()) {
+				long sleeping = world.getPlayers().stream().filter(player -> player.isSleeping() && canSleep(player)).count();
+				long active = world.getPlayers().stream().filter(Sleep::canSleep).count();
+				int needed = (int) Math.ceil(active / 2d);
+
+				if (sleeping >= needed && sleepingWorlds.get(world) != State.SKIPPING)
+					skipNight(world);
+				else if (sleepingWorlds.get(world) == State.SLEEPING)
+					world.getPlayers().forEach(player -> ActionBarUtils.sendActionBar(player,
+						"Sleepers needed to skip night: &e" + sleeping + "&3/&e" + needed));
 			}
-
-			Tasks.wait(wait, () -> {
-				world.setTime(0);
-				if (world.hasStorm())
-					world.setStorm(false);
-				if (world.isThundering())
-					world.setThundering(false);
-
-				handling = false;
-				lastCalculatedSleeping = 0;
-				lastCalculatedNeeded = 0;
-			});
-		}
+		});
 	}
 
-	protected boolean canSleep(Player player) {
+	private static boolean canSleep(Player player) {
 		return !PlayerUtils.isVanished(player) && !AFK.get(player).isTimeAfk() && player.getGameMode() == GameMode.SURVIVAL;
 	}
 
-	protected void handle(World world) {
-		if (!handling)
-			Tasks.wait(1, () -> calculate(world));
-	}
+	private static void skipNight(World world) {
+		sleepingWorlds.put(world, State.SKIPPING);
+		world.getPlayers().forEach(
+			player -> PlayerUtils.send(player, PREFIX + "The night was skipped because 50% of players slept!"));
 
-	protected void handle(HasPlayer player) {
-		handle(player.getPlayer().getWorld());
-	}
+		world.setStorm(false);
+		world.setThundering(false);
 
-	@EventHandler
-	public void onBedEnter(PlayerBedEnterEvent event) {
-		handle(event);
-	}
+		int emptyActionbarTaskId = Tasks.repeatAsync(0, 1, () ->
+			world.getPlayers().forEach(player -> ActionBarUtils.sendActionBar(player, " ")
+		));
+		int wait = 0;
+		while (true) {
+			long newTime = world.getTime() + (++wait * SPEED);
+			if (!new NumberRange(12541L, (24000L - SPEED)).containsNumber(newTime))
+				break;
 
-	@EventHandler
-	public void onBedLeave(PlayerBedLeaveEvent event) {
-		handle(event);
-	}
+			Tasks.wait(wait, () -> world.setTime(newTime));
+		}
 
-	@EventHandler
-	public void onPlayerQuit(PlayerQuitEvent event) {
-		handle(event);
+		Tasks.wait(wait, () -> {
+			world.setTime(0);
+			if (world.hasStorm())
+				world.setStorm(false);
+			if (world.isThundering())
+				world.setThundering(false);
+			Tasks.cancel(emptyActionbarTaskId);
+			sleepingWorlds.remove(world);
+		});
 	}
 
 	@EventHandler
-	public void onPlayerVanish(VanishEvent event) {
-		handle(event);
-	}
+	public void onBedEnter(PlayerBedEnterEvent event) {
+		World world = event.getPlayer().getWorld();
 
-	@EventHandler
-	public void onPlayerChangeWorlds(PlayerChangedWorldEvent event) {
-		handle(event);
+		// Is it day time?
```

Comment:
I normally prefer to create methods to explain what the code is doing, instead of comments. I find that comments easily get out of date, and its better to let the code explain itself. So for example, 
```java
if (!isDayTime(world))
    return;
    
if (!doDaylightCycle(world)) 
    return;
```

This is just preference so you don't have to change it, just sharing my knowledge :slightly_smiling_face: 

(also fwiw; I picked this up from Uncle Bob, highly recommend [his lecture series](https://www.youtube.com/watch?v=7EmboKQH8lM&list=PLUxszVpqZTNShoypLQW9a4dEcffsoZT4k) on youtube)

Reply:
I've heard a lot about Uncle Bob actually; been meaning to get around to watching/reading his stuff. Thanks for the recommendation!

And yeah, updated it. I always thought comments look ugly anyway lol